// react-hls-player/src/components/react-hls-player.js
import React from 'react';
import PropTypes from 'prop-types';
import Hls from 'hls.js';

class ReactHls extends React.Component {
  constructor (props) {
    super(props);

    this.state = {
      playerId : Date.now()
    };

    this.hls = null;
    this.video = React.createRef();
  }

  componentDidMount () {
    this._initPlayer();
  }

  componentWillUnmount () {
    if (this.hls) {
      this.hls.destroy();
    }
  }

  _initPlayer () {
    if (this.hls) {
      this.hls.destroy();
    }

    let { url, autoplay, hlsConfig } = this.props;
    let hls = new Hls(hlsConfig);

    hls.attachMedia(this.video.current);
    hls.on(Hls.Events.MEDIA_ATTACHED, () => {
      hls.loadSource(url);

      hls.on(Hls.Events.MANIFEST_PARSED, () => {
        if (autoplay) {
          this.video.current.play();
        }
      })
    })

    hls.on(Hls.Events.ERROR, function (event, data) {
      if (data.fatal) {
        switch(data.type) {
        case Hls.ErrorTypes.NETWORK_ERROR:
          hls.startLoad();
          break;
        case Hls.ErrorTypes.MEDIA_ERROR:
          hls.recoverMediaError();
          break;
        default:
          this._initPlayer();
          break;
        }
      }
    });

    this.hls = hls;
  }

  render () {
    let { playerId } = this.state;
    const { controls, width, height, poster, videoProps } = this.props;

    return (
      <div key={playerId} className="player-area">
        <video ref={this.video}
          className="hls-player"
          id={`react-hls-${playerId}`}
          controls={controls}
          width={width}
          height={height}
          poster={poster}
          {...videoProps}></video>
      </div>
    )
  }
}

ReactHls.propTypes = {
  url : PropTypes.string.isRequired,
  autoplay : PropTypes.bool,
  hlsConfig : PropTypes.object, //https://github.com/video-dev/hls.js/blob/master/docs/API.md#fine-tuning
  controls : PropTypes.bool,
  width : PropTypes.number,
  height : PropTypes.number,
  poster : PropTypes.string,
  videoProps : PropTypes.object
}

ReactHls.defaultProps = {
  autoplay : false,
  hlsConfig : {},
  controls : true,
  width : 500,
  height : 375
}

export default ReactHls;
